src/audio/module_adapter/module/volume/volume_generic_with_peakvol.c

```diff
@@ -0,0 +1,235 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2018 Intel Corporation. All rights reserved.
+//
+// Author: Liam Girdwood <liam.r.girdwood@linux.intel.com>
+//         Keyon Jie <yang.jie@linux.intel.com>
+//         Tomasz Lauda <tomasz.lauda@linux.intel.com>
+
+/**
+ * \file
+ * \brief Volume generic processing implementation
+ * \authors Liam Girdwood <liam.r.girdwood@linux.intel.com>\n
+ *          Keyon Jie <yang.jie@linux.intel.com>\n
+ *          Tomasz Lauda <tomasz.lauda@linux.intel.com>
+ */
+
+#include <sof/audio/buffer.h>
+#include <sof/audio/component.h>
+#include <sof/audio/format.h>
+#include <sof/common.h>
+#include <ipc/stream.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+LOG_MODULE_DECLARE(volume_generic, CONFIG_SOF_LOG_LEVEL);
+
+#include <sof/audio/volume.h>
+
+#ifdef CONFIG_GENERIC
+
+#if CONFIG_COMP_PEAK_VOL
+
+#if CONFIG_FORMAT_S24LE
+/**
+ * \brief Volume s24 to s24 multiply function
+ * \param[in] x   input sample.
+ * \param[in] vol gain.
+ * \return output sample.
+ *
+ * Volume multiply for 24 bit input and 24 bit output.
+ */
+static inline int32_t vol_mult_s24_to_s24(int32_t x, int32_t vol)
+{
+	return q_multsr_sat_32x32_24(sign_extend_s24(x), vol, Q_SHIFT_BITS_64(23, VOL_QXY_Y, 23));
+}
+
+/**
+ * \brief Volume processing from 24/32 bit to 24/32 bit.
+ * \param[in,out] dev Volume base component device.
+ * \param[in,out] sink Destination buffer.
+ * \param[in,out] source Input buffer.
+ * \param[in] frames Number of frames to process.
+ *
+ * Copy and scale volume from 24/32 bit source buffer
+ * to 24/32 bit destination buffer.
+ */
+static void vol_s24_to_s24(struct processing_module *mod, struct input_stream_buffer *bsource,
+			   struct output_stream_buffer *bsink, uint32_t frames)
+{
+	struct vol_data *cd = module_get_private_data(mod);
+	struct audio_stream __sparse_cache *source = bsource->data;
+	struct audio_stream __sparse_cache *sink = bsink->data;
+	int32_t vol;
+	int32_t *x, *x0;
+	int32_t *y, *y0;
+	int nmax, n, i, j;
+	const int nch = source->channels;
+	int remaining_samples = frames * nch;
+	int32_t tmp;
+
+	memset(cd->peak_regs.peak_meter, 0, sizeof(uint32_t) * cd->channels);
+	x = source->r_ptr;
+	y = sink->w_ptr;
+
+	bsource->consumed += VOL_S32_SAMPLES_TO_BYTES(remaining_samples);
+	bsink->size += VOL_S32_SAMPLES_TO_BYTES(remaining_samples);
+	while (remaining_samples) {
+		nmax = audio_stream_samples_without_wrap_s24(source, x);
+		n = MIN(remaining_samples, nmax);
+		nmax = audio_stream_samples_without_wrap_s24(sink, y);
+		n = MIN(n, nmax);
+		for (j = 0; j < nch; j++) {
+			x0 = x + j;
+			y0 = y + j;
+			vol = cd->volume[j];
+			tmp = 0;
+			for (i = 0; i < n; i += nch) {
+				y0[i] = vol_mult_s24_to_s24(x0[i], vol);
+				tmp = MAX(abs(*x0), tmp);
```

Comment:
before it was `tmp = MAX(*y0, tmp);` - are you sure you need `x0` and not `y0`? Was that a bug? Besides now you need `y0[i]` (or `x0[i]`) and not `*y0` - please check other places too.

Reply:
yes, I need the x0[i], since the peak volume of close source firmware is the maximum of input.